controllers/accounts_test.go

```diff
@@ -259,6 +269,183 @@ func (suite *AccountsTestSuite) TestAccountDeletionEndpointEnabled() {
+	suite.Nil(settings.RecoveryKeyCreatedAt)
+
+	// Enable 2FA
+	err := suite.ds.SetTOTPSetting(account.ID, true)
+	suite.Require().NoError(err)
+
+	// Test getting updated 2FA settings
+	req = httptest.NewRequest("GET", "/v2/accounts/2fa", nil)
+	req.Header.Set("Authorization", "Bearer "+token)
+	resp = util.ExecuteTestRequest(req, suite.router)
+	suite.Equal(http.StatusOK, resp.Code)
+
+	util.DecodeJSONTestResponse(suite.T(), resp.Body, &settings)
+	suite.True(settings.TOTP)
+	suite.NotNil(settings.TOTPEnabledAt)
+	suite.Nil(settings.RecoveryKeyCreatedAt)
+}
+
+func (suite *AccountsTestSuite) TestTOTPSetupAndFinalize() {
+	token, account := suite.createAuthSession()
+
+	// Test initializing TOTP setup
+	initReq := util.CreateJSONTestRequest("/v2/accounts/2fa/totp/init", controllers.TwoFAInitRequest{
+		GenerateQR: true,
+	})
+	initReq.Header.Set("Authorization", "Bearer "+token)
+	initResp := util.ExecuteTestRequest(initReq, suite.router)
+	suite.Equal(http.StatusOK, initResp.Code)
+
+	var initParsedResp controllers.TwoFAInitResponse
+	util.DecodeJSONTestResponse(suite.T(), initResp.Body, &initParsedResp)
+	suite.Require().NotEmpty(initParsedResp.URL)
+	suite.Require().NotNil(initParsedResp.QRCode)
+	suite.NotEmpty(*initParsedResp.QRCode)
+
+	// Parse the TOTP URL to extract the secret
+	key, err := otp.NewKeyFromURL(initParsedResp.URL)
+	suite.Require().NoError(err)
+	suite.Equal("Brave Account", key.Issuer())
+	suite.Equal("test@example.com", key.AccountName())
+	suite.True(strings.HasPrefix(initParsedResp.URL, "otpauth://totp/Brave%20Account"))
+
+	// Test finalizing TOTP setup with invalid code
+	invalidCode := "000000"
+	finalizeReq := util.CreateJSONTestRequest("/v2/accounts/2fa/totp/finalize", controllers.TwoFAFinalizeRequest{
+		Code: invalidCode,
+	})
+	finalizeReq.Header.Set("Authorization", "Bearer "+token)
+	finalizeResp := util.ExecuteTestRequest(finalizeReq, suite.router)
+	suite.Equal(http.StatusBadRequest, finalizeResp.Code)
+	util.AssertErrorResponseCode(suite.T(), finalizeResp, util.ErrBadTOTPCode.Code)
+
+	// Test finalizing TOTP setup with valid code
+	validCode, err := totp.GenerateCode(key.Secret(), time.Now().UTC())
+	suite.Require().NoError(err)
+
+	finalizeReq = util.CreateJSONTestRequest("/v2/accounts/2fa/totp/finalize", controllers.TwoFAFinalizeRequest{
+		Code: validCode,
+	})
+	finalizeReq.Header.Set("Authorization", "Bearer "+token)
+	finalizeResp = util.ExecuteTestRequest(finalizeReq, suite.router)
+	suite.Equal(http.StatusOK, finalizeResp.Code)
+
+	var finalizeParsedResp controllers.TwoFAFinalizeResponse
+	util.DecodeJSONTestResponse(suite.T(), finalizeResp.Body, &finalizeParsedResp)
+	suite.Require().NotNil(finalizeParsedResp.RecoveryKey)
+	suite.NotEmpty(*finalizeParsedResp.RecoveryKey)
+
+	// Verify 2FA is now enabled
+	updatedAccount, err := suite.ds.GetAccount(nil, account.Email)
+	suite.Require().NoError(err)
+	suite.True(updatedAccount.IsTwoFAEnabled())
+	suite.NotNil(updatedAccount.RecoveryKeyHash)
```

Comment:
We could actually hash the recovery in this test to check that it's right, though maybe it's not that valuable a test given that we would basically use the same function as the real code.

Reply:
it wouldn't hurt